File: _NPCScan.Tools/Tools/UpdateNPCScanOptions.lua
--[[ _NPCScan.Tools by Saiket
Tools/UpdateNPCScanOptions.lua - Updates a complete _NPCScan options file.

1. Find the latest versions of these DBC files in WoW's MPQ archives using a
   tool such as WinMPQ:
   * <DBFilesClient/Achievement.dbc>
   * <DBFilesClient/Achievement_Criteria.dbc>
   * <DBFilesClient/AreaTable.dbc>
   * <DBFilesClient/Map.dbc>
   * <DBFilesClient/WorldMapArea.dbc>
   * <DBFilesClient/WorldMapContinent.dbc>
2. Extract them to the <DBFilesClient> folder.
3. Run <DBCUtil.bat> to convert all found *.DBC files into *.CSV files using
   nneonneo's <DBCUtil.exe> program.

Once data files are ready, run this script with a standalone Lua 5.1 interpreter.
The <../_NPCScan.lua> data file will be overwritten.
]]

local OutputFilename = [[../_NPCScan.lua]];
local LEVEL_MAX = 85; -- Max rare level to query for

package.cpath = [[.\Libs\?.dll;]]..package.cpath;
package.path = [[.\Libs\?.lua;]]..package.path;
local http = require( "socket.http" );
require( "json" );
require( "bit" );
require( "DbcCsv" );




--- Doesn't allocate error messages until needed.
local function assertf ( Success, Format, ... )
	if ( Success ) then
		return Success;
	end
	local Args = { ... }; -- Convert all to strings
	for Index = 1, select( "#", ... ) do
		Args[ Index ] = tostring( Args[ Index ] );
	end
	error( Format:format( unpack( Args, 1, select( "#", ... ) ) ) );
end




-- Get list of all rares from WowHead
print( "Retrieving list of rares..." );
local NPCData = {};
do
	--- Adds query results to list, or recursively subdivides query if results omitted.
	local function QueryLevelRange ( LevelMin, LevelMax )
		print( ( "\tQuerying levels %d - %d..." ):format( LevelMin, LevelMax ) );
		local Path = ( [[http://www.wowhead.com/npcs?filter=cl=4:2;minle=%d;maxle=%d]] ):format(
			LevelMin, LevelMax );
		local Text, Status = http.request( Path );
		assertf( Text and Status == 200, "Request failed: Status code %d.", Status );

		local ListData = assert( Text:match(
			[[<script type="text/javascript">//<!%[CDATA%[%s*new Listview%((.-%));%s*//%]%]></script>]]
			), "Listview data not found." );
		local ResultsMax, ResultsShown = ListData:match( [[note: $WH%.sprintf%(LANG%.lvnote_npcsfound, (%d+), (%d+)%)]] );
		if ( ResultsMax ) then -- Too many results; Subdivide query
			print( ( "\t\tSubdividing %d query results." ):format( ResultsMax ) );
			local LevelMid = math.floor( ( LevelMin + LevelMax ) / 2 );
			QueryLevelRange( LevelMin, LevelMid );
			QueryLevelRange( LevelMid + 1, LevelMax );
		else -- Parse results
			local Data = json.decode( ListData );
			assertf( Data.template == "npc" and Data.id == "npcs",
				"Invalid list template %q or id %q.", Data.template, Data.id );
			for _, NPC in ipairs( assert( Data.data, "List data not found." ) ) do
				NPCData[ NPC.id ] = {
					Name = NPC.name;
					AreaIDs = ( NPC.location and #NPC.location > 0 and NPC.location ) or nil;
				};
			end
		end
	end
	QueryLevelRange( 1, LEVEL_MAX );
end


-- Remove achievement rares
print( "Removing achievement rares..." );
local AchievementsActive = {
	[ 1312 ] = true; -- Bloody Rare
	[ 2257 ] = true; -- Frostbitten
};
do
	local Achievements = DbcCsv.Parse( [[DBFilesClient/Achievement.dbc.csv]], 1,
		"ID", nil, nil, "CriteriaParent", "Name" );
	local AchievementFilter = {};

	--- Adds AchievementID and all its parents to achievement filter.
	local function AddAchievement ( AchievementID )
		AchievementFilter[ AchievementID ] = true;

		local Achievement = Achievements[ AchievementID ];
		print( ( "\tRemoving achievement [ %d ] %s." ):format(
			AchievementID, Achievement.Name ) );
		-- Recurse any achievements whos criteria must also be met
		if ( Achievement.CriteriaParent ~= 0 ) then
			return AddAchievement( Achievement.CriteriaParent );
		end
	end
	for AchievementID in pairs( AchievementsActive ) do
		AddAchievement( AchievementID );
	end

	local AchievementCriteria = DbcCsv.Parse( [[DBFilesClient/Achievement_Criteria.dbc.csv]], 1,
		"ID", "AchievementID", "Type", "AssetID" );
	for CriteriaID, Criteria in pairs( AchievementCriteria ) do
		if ( AchievementFilter[ Criteria.AchievementID ]
			and Criteria.Type == 0 -- Mob kill type
		) then
			NPCData[ Criteria.AssetID ] = nil;
		end
	end
end


-- Find continent order numbers for numeric WorldIDs
print( "Generating WorldIDs..." );
local WorldMapContinent = DbcCsv.Parse( [[DBFilesClient/WorldMapContinent.dbc.csv]], 2,
	"ID", "MapID" );
do
	local WorldMapAreas = DbcCsv.Parse( [[DBFilesClient/WorldMapArea.dbc.csv]], 1,
		"ID", "MapID", "AreaID", nil, nil, nil, nil, nil, nil, nil, nil, "Flags" );

	local Order = {};
	local FLAG_PHASE = 0x2;
	for ID, WorldMapArea in pairs( WorldMapAreas ) do
		local Continent = WorldMapContinent[ WorldMapArea.MapID ];
		if ( Continent and WorldMapArea.AreaID == 0 -- Is a continent
			and bit.band( tonumber( WorldMapArea.Flags, 16 ), FLAG_PHASE ) == 0 -- Not a phased map
		) then
			Continent.WorldMapAreaID, Order[ #Order + 1 ] = ID, Continent;
		end
	end
	-- Continents appear to be sorted by their WorldMapAreaIDs rather than ContinentIDs.
	-- Kalimdor and Eastern Kingdoms are in reverse order in WorldMapContinent.dbc.
	table.sort( Order, function ( Continent1, Continent2 )
		return Continent1.WorldMapAreaID < Continent2.WorldMapAreaID;
	end );
	for Index, Continent in ipairs( Order ) do
		Continent.Index = Index;
	end
end


local GetAreaContinent, GetAreaName;
do
	local Maps = DbcCsv.Parse( [[DBFilesClient/Map.dbc.csv]], 1,
		"ID", nil, nil, nil, nil, nil, "Name", "AreaID" );
	local AreaTable = DbcCsv.Parse( [[DBFilesClient/AreaTable.dbc.csv]], 1,
		"ID", "MapID", "ParentID", nil, nil, nil, nil, nil, nil, nil, nil, "Name" );

	--- @return Map table for given AreaID.
	local function GetAreaMap ( AreaID )
		return Maps[ AreaTable[ AreaID ].MapID ];
	end
	--- @return The continent ID of AreaID's map, or nil if not a continent.
	function GetAreaContinent ( AreaID )
		local Continent = WorldMapContinent[ GetAreaMap( AreaID ).ID ];
		return Continent and Continent.Index;
	end
	--- @return The localized name of AreaID's map.
	function GetAreaName ( AreaID )
		local Map = GetAreaMap( AreaID );
		if ( Map.AreaID == 0 ) then
			return Map.Name;
		else
			return AreaTable[ Map.AreaID ].Name;
		end
	end
end




-- Write _NPCScanOptionsCharacter data
print( "Writing _NPCScan options file..." );
local OutFile = assert( io.open( OutputFilename, "w+" ) );
assert( OutFile:write( [[
_NPCScanOptionsCharacter = {
	Version = "4.0.3.5";
	Achievements = {
]] ) );

-- Sort output by NPC ID
local Order = {};
for NpcID in pairs( NPCData ) do
	Order[ #Order + 1 ] = NpcID;
end
table.sort( Order );

-- Active achievement IDs
for AchievementID in pairs( AchievementsActive ) do
	assert( OutFile:write( ( "\t\t[ %d ] = true;\n" ):format( AchievementID ) ) );
end
assert( OutFile:write( [[
	};
	NPCs = {
]] ) );

-- Custom NPC names
for _, NpcID in ipairs( Order ) do
	assert( OutFile:write( ( "\t\t[ %d ] = %q;\n" ):format( NpcID, NPCData[ NpcID ].Name ) ) );
end
assert( OutFile:write( [[
	};
	NPCWorldIDs = {
]] ) );

-- Custom NPC worlds
for _, NpcID in ipairs( Order ) do
	local AreaIDs = NPCData[ NpcID ].AreaIDs;
	if ( AreaIDs ) then
		local WorldID;
		for _, AreaID in ipairs( AreaIDs ) do
			local WorldIDCurrent = GetAreaContinent( AreaID ) or GetAreaName( AreaID );
			if ( WorldID == nil ) then -- First world
				WorldID = WorldIDCurrent;
			elseif ( WorldID ~= WorldIDCurrent ) then -- Exists on multiple worlds
				WorldID = nil; -- Don't filter by world
				break;
			end
		end

		if ( WorldID ) then
			local Format = type( WorldID ) == "number"
				and "\t\t[ %d ] = %d;\n" or "\t\t[ %d ] = %q;\n";
			assert( OutFile:write( ( Format ):format( NpcID, WorldID ) ) );
		end
	end
end

assert( OutFile:write( [[
	};
};]] ) );
assert( OutFile:flush() );
assert( OutFile:close() );

File: _NPCScan.Tools/Update/UpdateNPCScanOptions.py
# -*- coding: utf_8 -*-
"""Generates a complete _NPCScan options file."""

import os.path

import wowdata.dbc as dbc
import wowdata.lua
import wowdata.mpq
import wowdata.wowhead

__author__ = 'Saiket'
__email__ = 'saiket.wow@gmail.com'
__license__ = 'GPL'

_KILL_ACHIEVEMENTS = set((  # Rare mob achievements tracked by _NPCScan
  1312,  # Bloody Rare
  2257,  # Frostbitten
))
_KILL_CRITERIA_TYPE = 0  # Criteria TypeID for NPC kills

def updateNPCScanOptions(dataPath, outputFilename, locale):
  """Writes a settings file including all rare mobs from Wowhead."""
  dataPath = os.path.normcase(dataPath)
  outputFilename = os.path.normcase(outputFilename)
  print 'Writing all rares from %s Wowhead to <%s>...' % (locale, outputFilename)
  npcs = wowdata.wowhead.getNPCsAllLevels(locale, cl='2:4')  # Rare and rare elite

  with wowdata.mpq.openLocaleMPQ(dataPath, locale) as archive:
    # Remove NPCs that _NPCScan already tracks as part of achievements
    print '\tRemoving achievement criteria rares...'
    with dbc.DBC(archive.open('DBFilesClient/Achievement.dbc'),
      'id', None, None, 'criteriaParent', 'name') as achievements \
    :
      achievements.rows = {achievement.int('id'): achievement for achievement in achievements}
      killAchievements = set()
      def addKillAchievement(achievementID):
        """Adds an achievement and all its parents to the filter."""
        achievement = achievements.rows[achievementID]
        print '\t\tAchievement%d - %r' % (achievementID, achievement.str('name'))
        killAchievements.add(achievementID)
        parentID = achievement.int('criteriaParent')
        if parentID:
          addKillAchievement(parentID)
      for achievementID in _KILL_ACHIEVEMENTS:
        addKillAchievement(achievementID)
    with dbc.DBC(archive.open('DBFilesClient/Achievement_Criteria.dbc'),
      'id', 'achievementID', 'type', 'assetID') as achievementCriteria \
    :
      for criteria in achievementCriteria:
        if (criteria.int('achievementID') in killAchievements
          and criteria.int('type') == _KILL_CRITERIA_TYPE
        ):
          npcID = criteria.int('assetID')
          if npcID in npcs:
            del npcs[npcID]

    # Generate ContinentIDs that match in-game APIs
    with dbc.DBC(archive.open('DBFilesClient/WorldMapArea.dbc'),
      'id', 'mapID', 'areaID', flags=11) as worldMapAreas \
    :
      worldMapAreas.rows = {worldMap.int('id'): worldMap for worldMap in worldMapAreas}
      FLAG_PHASE = 0x2
      mapContinentIDs = {mapID: continentID
        for continentID, mapID in enumerate(start=1, sequence=(worldMap.int('mapID')
          for worldMapID, worldMap in sorted(worldMapAreas.rows.iteritems())  # Ordered by WorldMapID
            if not worldMap.int('areaID')  # Is a continent
              and not worldMap.int('flags') & FLAG_PHASE))}  # Not a phased map

    with \
      dbc.DBC(archive.open('DBFilesClient/AreaTable.dbc'),
        'id', 'mapID', 'parentID', name=11) as areaTable, \
      dbc.DBC(archive.open('DBFilesClient/Map.dbc'),
        'id', name=6, areaID=7) as maps \
    :
      areaTable.rows = {area.int('id'): area for area in areaTable}
      maps.rows = {map.int('id'): map for map in maps}

      def getAreaWorldID(areaID):
        """Returns a "WorldID" for the map containing the given area ID."""
        mapID = areaTable.rows[areaID].int('mapID')
        if mapID in mapContinentIDs:
          return mapContinentIDs[mapID]
        else:  # Use localized name
          areaID = maps.rows[mapID].int('areaID')
          row = areaTable.rows[areaID] if areaID else maps.rows[mapID]
          return row.str('name')

      with open(outputFilename, 'w+b') as output:
        output.write('_NPCScanOptionsCharacter = {\n')
        output.write('\tVersion = "4.0.3.5";\n')
        output.write('\tAchievements = {\n')
        for achievementID in sorted(_KILL_ACHIEVEMENTS):
          output.write('\t\t[ %d ] = true;\n' % achievementID)
        output.write('\t};\n')
        output.write('\tNPCs = {\n')
        for npcID, npcData in sorted(npcs.iteritems()):
          output.write('\t\t[ %d ] = %s;\n' % (npcID, wowdata.lua.escapeData(npcData['name'])))
        output.write('\t};\n')
        output.write('\tNPCWorldIDs = {\n')
        for npcID, npcData in sorted(npcs.iteritems()):
          if 'location' in npcData:
            worldIDs = set(getAreaWorldID(areaID) for areaID in npcData['location'])
            if len(worldIDs) == 1:  # Only spawns on one map
              worldID = worldIDs.pop()
              if isinstance(worldID, unicode):
                output.write('\t\t[ %d ] = %s;\n' % (
                  npcID, wowdata.lua.escapeData(worldID.encode('utf_8'))))
              else: # ContinentID
                output.write('\t\t[ %d ] = %d;\n' % (npcID, worldID))
        output.write('\t};\n')
        output.write('};')


if __name__ == '__main__':
  import argparse
  parser = argparse.ArgumentParser(
    description='Generates a complete _NPCScan options file including all know rares.')
  parser.add_argument('--locale', '-l', type=unicode, required=True,
    help='Locale code to retrieve data for.')
  parser.add_argument('dataPath', type=unicode,
    help='The path to World of Warcraft\'s Data folder.')
  parser.add_argument('outputFilename', type=unicode,
    help='Output path for the resulting Lua source file.')
  updateNPCScanOptions(**vars(parser.parse_args()))